### PR TITLE
feat(logger): use AsyncLogger instead of g3log

### PR DESCRIPTION
**Pull request by Yimura**
Uses [Yimura/AsyncLogger](https://github.com/Yimura/AsyncLogger), it's based on g3log's implementation of logging except with a few touches of C++20.

This is a draft and I still need to implement stack dumping and exception recovery.

Closes #612

### CODE REVIEW

**Comment by GeopJr** [comment]
AsyncLogger needs a license

**Comment by Yimura** [comment]
> AsyncLogger needs a license

Completely forgot about that, thanks for the heads up.

**Comment by Yimura** [comment]
Haven't added a stack dumper yet, also need to take a look at how we should handle a waterfall of exceptions.

**Comment by Yimura** [comment]
Works fine except for 2 issues I'm having:
1. Doesn't load symbols
2. Crashes after logging stack to console

**Comment by Yimura** [comment]
Maybe remove the module list since there are no differences anymore between game versions?

Example stack trace:
```
[22:05:03.5215624][ERROR/exception_handler.cpp:26] Exception Received
Dumping modules:
GTA5.exe Base Address: 0x7FF730450000 Size: 63296000
ntdll.dll Base Address: 0x7FFA3F2E0000 Size: 1224704
KERNEL32.DLL Base Address: 0x7FFA3D220000 Size: 516096
KERNELBASE.dll Base Address: 0x7FFA3CBD0000 Size: 1536000
USER32.dll Base Address: 0x7FFA3E300000 Size: 610304
win32u.dll Base Address: 0x7FFA3CA00000 Size: 49152
GDI32.dll Base Address: 0x7FFA3D5F0000 Size: 57344
gdi32full.dll Base Address: 0x7FFA3CA30000 Size: 684032
msvcp_win.dll Base Address: 0x7FFA3CF50000 Size: 348160
ucrtbase.dll Base Address: 0x7FFA3C8E0000 Size: 798720
PSAPI.DLL Base Address: 0x7FFA3DBB0000 Size: 4096
WS2_32.dll Base Address: 0x7FFA3DB00000 Size: 286720
VERSION.dll Base Address: 0x7FF9EC360000 Size: 156672
RPCRT4.dll Base Address: 0x7FFA3D3F0000 Size: 913408
DSOUND.dll Base Address: 0x7FF9EC2B0000 Size: 491520
msvcrt.dll Base Address: 0x7FFA3D2E0000 Size: 495616
CRYPT32.dll Base Address: 0x7FFA3CFF0000 Size: 1126400
advapi32.dll Base Address: 0x7FFA3E8C0000 Size: 425984
WINTRUST.dll Base Address: 0x7FFA3C870000 Size: 299008
sechost.dll Base Address: 0x7FFA3DE70000 Size: 421888
IMM32.dll Base Address: 0x7FFA3E040000 Size: 131072
combase.dll Base Address: 0x7FFA3E4B0000 Size: 2490368
SHLWAPI.dll Base Address: 0x7FFA3D390000 Size: 212992
SHELL32.dll Base Address: 0x7FFA3E970000 Size: 6287360
WININET.dll Base Address: 0x7FFA2C840000 Size: 2170880
ole32.dll Base Address: 0x7FFA3E080000 Size: 892928
OLEAUT32.dll Base Address: 0x7FFA3D510000 Size: 643072
WINMM.dll Base Address: 0x7FFA36300000 Size: 114688
MFPlat.DLL Base Address: 0x7FFA36B90000 Size: 1503232
MF.dll Base Address: 0x7FFA22DC0000 Size: 368640
shcore.dll Base Address: 0x7FFA3F150000 Size: 663552
msdmo.dll Base Address: 0x7FF9DE810000 Size: 16384
MFReadWrite.dll Base Address: 0x7FF9EA840000 Size: 999424
PROPSYS.dll Base Address: 0x7FFA38660000 Size: 651264
ResampleDmo.DLL Base Address: 0x7FF9D82A0000 Size: 503808
WTSAPI32.dll Base Address: 0x7FFA38B90000 Size: 40960
d3d9.dll Base Address: 0x7FFA12E50000 Size: 1294336
D3DCOMPILER_43.dll Base Address: 0x7FF9D78C0000 Size: 2423296
powrprof.dll Base Address: 0x7FFA3C0D0000 Size: 77824
DINPUT8.dll Base Address: 0x7FF9EC550000 Size: 176128
XINPUT1_3.dll Base Address: 0x400000 Size: 85504
SETUPAPI.dll Base Address: 0x7FFA3D620000 Size: 892928
bcrypt.dll Base Address: 0x7FFA3C190000 Size: 106496
IPHLPAPI.DLL Base Address: 0x7FFA3B360000 Size: 122880
MSVCP140.dll Base Address: 0x7FF9E7A60000 Size: 346112
VCRUNTIME140.dll Base Address: 0x7FF9E7AF0000 Size: 65024
VCRUNTIME140_1.dll Base Address: 0x7FF9E8C30000 Size: 15360
winmmbase.dll Base Address: 0x7FFA2CEE0000 Size: 102400
cfgmgr32.dll Base Address: 0x7FFA3C480000 Size: 217088
kernel.appcore.dll Base Address: 0x7FFA3B7F0000 Size: 36864
dwmapi.dll Base Address: 0x7FFA3A0F0000 Size: 98304
dxcore.dll Base Address: 0x7FFA39C60000 Size: 143360
GFSDK_TXAA_AlphaResolve.win64.dll Base Address: 0x7FF9EE850000 Size: 48128
OPENGL32.dll Base Address: 0x7FF9EC6A0000 Size: 806912
bink2w64.dll Base Address: 0x7FF9DE2D0000 Size: 307712
GFSDK_ShadowLib.win64.dll Base Address: 0x7FF9D73A0000 Size: 96256
GLU32.dll Base Address: 0x7FF9EE800000 Size: 126976
MFCORE.DLL Base Address: 0x7FF9F63C0000 Size: 3207168
d3dx9_43.dll Base Address: 0x7FF9D62B0000 Size: 2280960
version.dll Base Address: 0x7FFA36840000 Size: 12288
UMPDC.dll Base Address: 0x7FFA3C0B0000 Size: 36864
bcryptPrimitives.dll Base Address: 0x7FFA3CB50000 Size: 397312
RTWorkQ.DLL Base Address: 0x7FFA36AA0000 Size: 102400
MSASN1.dll Base Address: 0x7FFA3C020000 Size: 36864
gameoverlayrenderer64.dll Base Address: 0x7FF9EE110000 Size: 982528
inputhost.dll Base Address: 0x7FFA24CD0000 Size: 1351680
CoreMessaging.dll Base Address: 0x7FFA395D0000 Size: 856064
CRYPTBASE.DLL Base Address: 0x7FFA3BE70000 Size: 12288
windows.storage.dll Base Address: 0x7FFA3A860000 Size: 6569984
wintypes.dll Base Address: 0x7FFA3A6F0000 Size: 540672
dbghelp.dll Base Address: 0x7FFA3A4A0000 Size: 1581056
dbgcore.DLL Base Address: 0x7FF9E7B60000 Size: 126976
uxtheme.dll Base Address: 0x7FFA39B40000 Size: 425984
clbcatq.dll Base Address: 0x7FFA3D160000 Size: 491520
amsi.dll Base Address: 0x7FFA1D1A0000 Size: 73728
USERENV.dll Base Address: 0x7FFA3BD20000 Size: 90112
profapi.dll Base Address: 0x7FFA3C6E0000 Size: 73728
MpOav.dll Base Address: 0x7FFA1D120000 Size: 204800
apphelp.dll Base Address: 0x7FFA399C0000 Size: 323584
CRYPTSP.dll Base Address: 0x7FFA3BF60000 Size: 49152
rsaenh.dll Base Address: 0x7FFA3B750000 Size: 147456
imagehlp.dll Base Address: 0x7FFA3F130000 Size: 49152
gpapi.dll Base Address: 0x7FFA3BCF0000 Size: 73728
cryptnet.dll Base Address: 0x7FFA234E0000 Size: 135168
WINNSI.DLL Base Address: 0x7FFA36870000 Size: 12288
NSI.dll Base Address: 0x7FFA3DF10000 Size: 8192
steam_api64.dll Base Address: 0x763E0000 Size: 110592
steamclient64.dll Base Address: 0x7FF9EC950000 Size: 15837184
vstdlib_s64.dll Base Address: 0x7FF9EE050000 Size: 285184
tier0_s64.dll Base Address: 0x7FF9EC7B0000 Size: 257536
Secur32.dll Base Address: 0x7FFA22C60000 Size: 12288
MSWSOCK.dll Base Address: 0x7FFA3BC80000 Size: 327680
SSPICLI.DLL Base Address: 0x7FFA3BA80000 Size: 126976
D3D11.DLL Base Address: 0x7FFA347B0000 Size: 1982464
dxgi.dll Base Address: 0x7FFA39CA0000 Size: 630784
directxdatabasehelper.dll Base Address: 0x7FFA36B30000 Size: 184320
ntmarta.dll Base Address: 0x7FFA3B810000 Size: 147456
DEVOBJ.dll Base Address: 0x7FFA3C4D0000 Size: 110592
D3D10_1.DLL Base Address: 0x7FF9D7880000 Size: 147456
d3d10_1core.dll Base Address: 0x7FF9DD610000 Size: 12288
MSCTF.dll Base Address: 0x7FFA3DF20000 Size: 913408
textinputframework.dll Base Address: 0x7FFA28230000 Size: 925696
WINSTA.dll Base Address: 0x7FFA3BFB0000 Size: 221184
aticfx64.dll Base Address: 0x7FFA31BC0000 Size: 1124864
atiuxp64.dll Base Address: 0x7FFA318C0000 Size: 190464
atidxx64.dll Base Address: 0x7FFA2F990000 Size: 11611648
amdihk64.dll Base Address: 0x7FFA2F2E0000 Size: 116736
D3DSCache.dll Base Address: 0x7FFA2EE70000 Size: 102400
twinapi.appcore.dll Base Address: 0x7FFA32B80000 Size: 1794048
DNSAPI.dll Base Address: 0x7FFA3B390000 Size: 696320
rasadhlp.dll Base Address: 0x7FFA2EEA0000 Size: 8192
fwpuclnt.dll Base Address: 0x7FFA33D60000 Size: 290816
schannel.DLL Base Address: 0x7FFA3B650000 Size: 544768
mskeyprotect.dll Base Address: 0x7FFA1C4F0000 Size: 45056
NTASN1.dll Base Address: 0x7FFA3C120000 Size: 69632
ncrypt.dll Base Address: 0x7FFA3C160000 Size: 102400
ncryptsslp.dll Base Address: 0x7FFA37140000 Size: 86016
HID.DLL Base Address: 0x7FFA3B0D0000 Size: 24576
MMDevApi.dll Base Address: 0x7FFA31B20000 Size: 462848
XAudio2_7.dll Base Address: 0x7FF9D7310000 Size: 485376
AUDIOSES.DLL Base Address: 0x7FFA2AEF0000 Size: 1318912
resourcepolicyclient.dll Base Address: 0x7FFA39DE0000 Size: 36864
Windows.UI.dll Base Address: 0x7FFA2ACF0000 Size: 983040
avrt.dll Base Address: 0x7FFA36830000 Size: 8192
napinsp.dll Base Address: 0x7FFA1B3F0000 Size: 53248
pnrpnsp.dll Base Address: 0x7FFA1BA40000 Size: 69632
winrnr.dll Base Address: 0x7FFA1BA60000 Size: 36864
wshbth.dll Base Address: 0x7FFA33D20000 Size: 45056
nlansp_c.dll Base Address: 0x7FFA1BD10000 Size: 73728
socialclub.dll Base Address: 0x7FF9DD680000 Size: 4881920
chrome_elf.dll Base Address: 0x7FF9EDF40000 Size: 685568
libcef.dll Base Address: 0x7FF9BCBB0000 Size: 109755392
UIAutomationCore.DLL Base Address: 0x7FFA10BF0000 Size: 3043328
COMDLG32.dll Base Address: 0x7FFA3DD80000 Size: 745472
NETAPI32.dll Base Address: 0x7FFA2D020000 Size: 49152
OLEACC.dll Base Address: 0x7FFA0FA10000 Size: 274432
COMCTL32.dll Base Address: 0x7FFA23760000 Size: 569344
DWrite.dll Base Address: 0x7FFA25910000 Size: 1589248
WINSPOOL.DRV Base Address: 0x7FFA1FC00000 Size: 368640
USP10.dll Base Address: 0x7FFA22C80000 Size: 4096
credui.dll Base Address: 0x7FFA33210000 Size: 65536
dxva2.dll Base Address: 0x7FFA23250000 Size: 102400
urlmon.dll Base Address: 0x7FFA35760000 Size: 1302528
CRYPTUI.dll Base Address: 0x7FFA22260000 Size: 282624
WINHTTP.dll Base Address: 0x7FFA36E00000 Size: 884736
dhcpcsvc.DLL Base Address: 0x7FFA37350000 Size: 65536
wevtapi.dll Base Address: 0x7FFA35C70000 Size: 258048
ESENT.dll Base Address: 0x7FFA1E5A0000 Size: 3117056
iertutil.dll Base Address: 0x7FFA35470000 Size: 569344
srvcli.dll Base Address: 0x7FFA35730000 Size: 57344
netutils.dll Base Address: 0x7FFA3B2E0000 Size: 20480
WKSCLI.DLL Base Address: 0x7FFA35A50000 Size: 49152
SAMCLI.DLL Base Address: 0x7FFA2CF10000 Size: 61440
FirewallAPI.dll Base Address: 0x7FFA3B230000 Size: 380928
fwbase.dll Base Address: 0x7FFA3B1F0000 Size: 151552
FWPolicyIOMgr.dll Base Address: 0x7FFA2EDA0000 Size: 184320
wdmaud.drv Base Address: 0x7FFA07F80000 Size: 200704
ksuser.dll Base Address: 0x7FFA08D70000 Size: 8192
msacm32.drv Base Address: 0x7FFA08D60000 Size: 20480
MSACM32.dll Base Address: 0x7FFA07F60000 Size: 77824
midimap.dll Base Address: 0x7FFA07F50000 Size: 16384
ondemandconnroutehelper.dll Base Address: 0x7FFA0CA30000 Size: 45056
xinput1_4.dll Base Address: 0x7FF9F6930000 Size: 32768
dhcpcsvc6.DLL Base Address: 0x7FFA37190000 Size: 28672
CoreUIComponents.dll Base Address: 0x7FFA35CD0000 Size: 2051584
YimMenu.dll Base Address: 0x7FF9BB950000 Size: 5312512
D3DCOMPILER_47.dll Base Address: 0x7FFA38D00000 Size: 3477504
Dumping registers:
RAX: 0x1
RCX: 0x70
RDX: 0xFFFFFFFA
RBX: 0x255BB19DA80
RSI: 0x2557E3EF5B0
RDI: 0x0
RSP: 0xD578FFFA00
RBP: 0xD578FFFB00
R8:  0x25538FCD428
R9:  0xD2
R10: 0x7FF9BBEC4918
R11: 0xD578FFF9A0
R12: 0x0
R13: 0x0
R14: 0x1340
R15: 0x0
Dumping stacktrace:
[0]     YimMenu.dll+0x2BE59F 0x7FF9BBC0E59F
[1]     YimMenu.dll+0x2B9F80 0x7FF9BBC09F80
[2]     YimMenu.dll+0x1807FC 0x7FF9BBAD07FC
[3]     YimMenu.dll+0x20AD2F 0x7FF9BBB5AD2F
[4]     YimMenu.dll+0x194109 0x7FF9BBAE4109
[5]     GTA5.exe+0x1368E4C 0x7FF7317B8E4C
[6]     GTA5.exe+0x1635809 0x7FF731A85809
[7]     GTA5.exe+0x12E93DA 0x7FF7317393DA
[8]     KERNEL32.DLL+0x155A0 0x7FFA3D2355A0
[9]     ntdll.dll+0x485B 0x7FFA3F2E485B
--------End of exception--------
```

**Comment by Yimura** [comment]
Alright I'm happy with it currently, module symbols aren't loaded for some reason but maybe we can fix that later on.

@tupoy-ya, could you check this out to see if it works on Linux? `Settings->Debug->MOV 0xdead` to test it.

File: src/views/debug/view_debug_misc.cpp [hunk 1]
```diff
@@ -15,6 +15,17 @@ namespace big
 	{
 		if (ImGui::BeginTabItem("DEBUG_TAB_MISC"_T.data()))
 		{
+			if (components::button("MOV QWORD"))
+			{
+				*static_cast<uint64_t*>(nullptr) = 0;
+				uint64_t i = *static_cast<uint64_t*>(nullptr);
+            }
```
Comment (Yimura) [inline src/views/debug/view_debug_misc.cpp]: Random weird indent

File: src/logger/stack_trace.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+namespace big
+{
+    stack_trace::stack_trace(EXCEPTION_POINTERS* exception_info) :
+        m_exception_info(exception_info),
+        m_frame_pointers(32)
+    {
+        static std::mutex m;
+        std::lock_guard lock(m);
+
+        SymInitialize(GetCurrentProcess(), nullptr, true);
```
Comment (Yimura) [inline src/logger/stack_trace.cpp:15]: Add `SymSetOptions(SYMOPT_LOAD_LINES)` for later testing

**Comment by xiaoxiao921** [comment]
I'd say keep the module list just because a crash may arise because of an incompatibility with another menu or even another software

**Comment by tupoy-ya** [comment]
> Alright I'm happy with it currently, module symbols aren't loaded for some reason but maybe we can fix that later on.
> 
> @tupoy-ya, could you check this out to see if it works on Linux? `Settings->Debug->MOV 0xdead` to test it.

<details>
  <summary>Crash log screenshot</summary>

image
</details>

Seems to work, but the log itself looks even worse then it did before.

<details>
  <summary>Before</summary>

image


</details>

<details>
  <summary>After</summary>

image


</details>

**Comment by Yimura** [comment]
> Seems to work, but the log itself looks even worse then it did before.

I guess you're talking about the colours not being parsed by wine as well as the timestamp including nanoseconds.
I will take a look at fixing those colours however I doubt it's possible since I'm using the correct API from Windows to enable colour output.

**Comment by tupoy-ya** [comment]
> > Seems to work, but the log itself looks even worse then it did before.
> 
> I guess you're talking about the colours not being parsed by wine as well as the timestamp including nanoseconds. I will take a look at fixing those colours however I doubt it's possible since I'm using the correct API from Windows to enable colour output.

Colors not being parsed is probably a `wine cmd` issue and i'm pretty sure it cannot be fixed, the problem is that earlier  text was separated from color tags by a space.

Maybe we can detect if we are running in wine and disable color tags in console.

**Comment by xiaoxiao921** [comment]
You can yeah, wine leave a process env variable iirc and most likely other things

**Comment by Yimura** [comment]
Alright I fixed symbols for the logger and implemented a method of detecting Wine hopefully it works.

@tupoy-ya, care to test if it successfully detects Wine?

**Comment by GeopJr** [comment]
You might as well support the other 'standards':

- `TERM` being set to `dumb`
- `NO_COLOR` being set at all (see https://no-color.org/)

**Comment by tupoy-ya** [comment]
Now it does not inject.

I will try to fix it tomorrow.

**Comment by Yimura** [comment]
> Now it does not inject.
>
> I will try to fix it tomorrow.

Nothing at all or is there an error that occurs on inject? (Like for example Cheat Engine shows if a DLL isn't found while calling LoadLibrary remotely)

**Comment by tupoy-ya** [comment]
> > Now it does not inject.
> > I will try to fix it tomorrow.
> 
> Nothing at all or is there an error that occurs on inject? (Like for example Cheat Engine shows if a DLL isn't found while calling LoadLibrary remotely)

Nothing at all, tried looking at wine logs but they do not seem to contain anything useful, or maybe i just didn't find it in 10mb of
```
fixme:keyboard:NtUserActivateKeyboardLayout flags 40000000 not supported
fixme:keyboard:X11DRV_ActivateKeyboardLayout 0x4090409, 40000000: semi-stub!
```

**Comment by tupoy-ya** [comment]
It seems that we cannot call `wine_get_version` or it will fail to start.
With this
```cpp
if (module.get_export("wine_get_version"))
        {
            LOG(VERBOSE) << "Wine environment detected, swapping to simple logger.";
            m_console_logger = &logger::format_console_simple;
        }
```
i get an exception, after which everything runs fine.

Also WineHQ says that doing wine specific patches is a bad idea, so maybe we should instead tell wine users to set `NO_COLOR` if they don't want broken color tags.
https://wiki.winehq.org/Developer_FAQ#How_can_I_detect_Wine.3F

**Comment by Yimura** [comment]
> It seems that we cannot call `wine_get_version` or it will fail to start. With this
> 
> ```c++
> if (module.get_export("wine_get_version"))
>         {
>             LOG(VERBOSE) << "Wine environment detected, swapping to simple logger.";
>             m_console_logger = &logger::format_console_simple;
>         }
> ```
> 
> i get an exception, after which everything runs fine.
> 
> Also WineHQ says that doing wine specific patches is a bad idea, so maybe we should instead tell wine users to set `NO_COLOR` if they don't want broken color tags. https://wiki.winehq.org/Developer_FAQ#How_can_I_detect_Wine.3F

Alternatively we just don’t call wine version? As we don’t really need to log that wine version to console, just logging something generalised would better.

**Comment by GeopJr** [comment]
On windows: give powershell a try too

**Comment by Yimura** [comment]
This is ready to be merged, waiting a bit before merging.

**Pull request merged**